sklearn/impute.py

```diff
@@ -271,7 +314,10 @@ def _dense_fit(self, X, strategy, missing_values):
             X = X.transpose()
             mask = mask.transpose()
 
-            most_frequent = np.empty(X.shape[0])
+            if X.dtype.kind == "O":
+                most_frequent = np.empty(X.shape[0], dtype=object)
+            else:
+                most_frequent = np.empty(X.shape[0])
```

Comment:
Maybe this should be:

```python
most_frequent = np.empty(X.shape[0], dtype=X.dtype)
```

Reply:
so discussed here: if you have integer with all -1, specify missing_value of -1, then the most_frequent will be NaN (due to implementation details to signal this case), and then X.dtype would clash with NaN.